sdk/lib/udmihelp/udmihelp.c

```diff
@@ -0,0 +1,166 @@
+    if (pGetSystemFirmwareTable)
+    {
+        dwBytes = pGetSystemFirmwareTable('RSMB', 0, NULL, 0);
+        if (dwBytes > 0)
+        {
+            pBuffer = malloc(dwBytes);
+            if (!pBuffer)
+            {
+                return NULL;
+            }
+            pGetSystemFirmwareTable('RSMB', 0, pBuffer, dwBytes);
+        }
+    }
+    if (dwBytes == 0)
+    {
+        // try use registry (works on NT 5.1)
+        if (RegOpenKeyW(HKEY_LOCAL_MACHINE,
+                        L"SYSTEM\\CurrentControlSet\\Services\\mssmbios\\Data",
+                        &hKey) != ERROR_SUCCESS)
+        {
+            return NULL;
+        }
+
+        if (RegQueryValueExW(hKey,
+                             L"SMBiosData",
+                             NULL,
+                             &dwType,
+                             NULL,
+                             &dwBytes) != ERROR_SUCCESS || dwType != REG_BINARY)
+        {
+            RegCloseKey(hKey);
+            return NULL;
+        }
+
+        pBuffer = malloc(dwBytes);
+        if (!pBuffer)
+        {
+            RegCloseKey(hKey);
+            return NULL;
+        }
+
+        if (RegQueryValueExW(hKey,
+                             L"SMBiosData",
+                             NULL,
+                             &dwType,
+                             pBuffer,
+                             &dwBytes) != ERROR_SUCCESS || dwType != REG_BINARY)
+        {
+            free(pBuffer);
+            RegCloseKey(hKey);
+            return NULL;
+        }
+
+        RegCloseKey(hKey);
+    }
+    ParseSMBiosTables(pBuffer, dwBytes, Strings);
+    return pBuffer;
+}
+
+/* Trim converted DMI string */
+VOID
+TrimDmiStringW(LPWSTR lpStr)
+{
+    unsigned Length;
+    int i = 0;
+
+    if (!lpStr)
+    {
+        return;
+    }
+
+    Length = wcslen(lpStr);
+    if (Length == 0)
+    {
+        return;
+    }
+
+    /* Trim leading spaces */
+    while (i < Length && lpStr[i] <= 32)
+    {
+        i++;
+    }
+    if (i < Length)
+    {
+        if (i > 0)
+        {
+            Length -= i;
+            memcpy(lpStr, &lpStr[i], (Length + 1) * sizeof(WCHAR));
+        }
+
+        /* Trim trailing spaces */
+        while (Length && lpStr[Length-1] <= 32)
+        {
+            lpStr[Length-1] = L'\0';
+            --Length;
+        }
+    }
+    else
+    {
+        *lpStr = 0;
+    }
+}
+
+/* Convert string from SMBIOS */
+unsigned
+GetSMBiosStringW(PCHAR DmiString, LPWSTR lpBuf, DWORD cchBuf, BOOL bTrim)
+{
+    unsigned cChars;
+
+    if (!DmiString)
+    {
+        return 0;
+    }
+
+    cChars = strlen(DmiString);
+    MultiByteToWideChar(CP_OEMCP, 0, DmiString, -1, lpBuf, cchBuf);
+
+    /* NULL-terminate string */
+    lpBuf[min(cchBuf-1, cChars)] = L'\0';
+
+    if (bTrim)
+    {
+        TrimDmiStringW(lpBuf);
+    }
+
+    return wcslen(lpBuf);
+}
+
+/* Free SMBIOS Data */
+VOID
+FreeSMBiosData(PVOID Buffer)
```

Comment:
Very good!
Memory allocated in a library should always be freed from that same library 👍